### PR TITLE
Add support for WASM Global Variables

### PR DESCRIPTION
**Pull request by akashin**
This PR implements support for global variable loading. The necessary steps are:
1. Generate globals and their initial values in the preamble of ZK ASM
2. In CLIF, generate accesses to globals as memory loads/stores at special "Globals" Symbol + offset which is the index of the global
3. In ISLE, detect accesses to this special "Globals" Symbol and replace them with a Global AMode
4. In Inst::Load and Inst::Store handle the Global AMode with appropriate ZKASM `MLOAD` and `MSTORE`

### CODE REVIEW

File: cranelift/codegen/src/isa/zkasm/lower/isle.rs [hunk 1]
```diff
@@ -364,6 +364,20 @@ impl generated_code::Context for ZkAsmIsleContext<'_, '_, MInst, ZkAsmBackend> {
     fn store_op(&mut self, ty: Type) -> StoreOP {
         StoreOP::from_type(ty)
     }
+
+    fn is_global_base(&mut self, name: ExternalName) -> Option<ExternalName> {
+        match name {
+            ExternalName::User(user_name_ref) => {
+                if user_name_ref.as_u32() == 1 {
```
Comment (akashin) [inline cranelift/codegen/src/isa/zkasm/lower/isle.rs:371]: Ideally this check would be more robust and less coupled with the environment logic...

File: cranelift/codegen/src/isa/zkasm/lower/isle.rs [hunk 2]
```diff
@@ -374,6 +388,12 @@ impl generated_code::Context for ZkAsmIsleContext<'_, '_, MInst, ZkAsmBackend> {
         tmp.to_reg()
     }
 
+    fn load_global_base(&mut self) -> Reg {
```
Comment (akashin) [inline cranelift/codegen/src/isa/zkasm/lower/isle.rs:391]: This probably needs to be replaced with a marker rule that doesn't have a direct lowering but is instead matched in the AMode generating ISLE rule

**Comment by MCJOHN974** [comment]
LGTM!

**Review by MCJOHN974 — APPROVED** [review]

**Comment by MCJOHN974** [comment]
Also, what is performance difference between globals and memory in zkasm? maybe we better should reserve some dedicated memory region for globals, will it faster? (I guess no, but sometimes zkasm provide some surprises)

**Pull request merged**